# PyCity Schools

### Dependencies and Starter Code

In [441]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load (Remember to Change These)
school_data_to_load = "Resources/schools_complete.csv"
student_data_to_load = "Resources/students_complete.csv"

# Read School and Student Data File and store into Pandas DataFrames
school_data = pd.read_csv(school_data_to_load)
student_data = pd.read_csv(student_data_to_load)

# Combine the data into a single dataset.  
school_data_complete = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])

### Check School Data types:

In [442]:
school_data_complete.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 39170 entries, 0 to 39169
Data columns (total 11 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Student ID     39170 non-null  int64 
 1   student_name   39170 non-null  object
 2   gender         39170 non-null  object
 3   grade          39170 non-null  object
 4   school_name    39170 non-null  object
 5   reading_score  39170 non-null  int64 
 6   math_score     39170 non-null  int64 
 7   School ID      39170 non-null  int64 
 8   type           39170 non-null  object
 9   size           39170 non-null  int64 
 10  budget         39170 non-null  int64 
dtypes: int64(6), object(5)
memory usage: 3.6+ MB


### Displaying merged DataFrame

In [443]:
school_data_complete.head()

Student ID       student_name gender grade        school_name  \
0           0       Paul Bradley      M   9th  Huang High School   
1           1       Victor Smith      M  12th  Huang High School   
2           2    Kevin Rodriguez      M  12th  Huang High School   
3           3  Dr. Richard Scott      M  12th  Huang High School   
4           4         Bonnie Ray      F   9th  Huang High School   

   reading_score  math_score  School ID      type  size   budget  
0             66          79          0  District  2917  1910635  
1             94          61          0  District  2917  1910635  
2             90          60          0  District  2917  1910635  
3             67          58          0  District  2917  1910635  
4             97          84          0  District  2917  1910635

# District Summary

* Calculate the total number of schools

* Calculate the total number of students

* Calculate the total budget

* Calculate the average math score 

* Calculate the average reading score

* Calculate the percentage of students with a passing math score (70 or greater)

* Calculate the percentage of students with a passing reading score (70 or greater)

* Calculate the percentage of students who passed math **and** reading (% Overall Passing)

* Create a dataframe to hold the above results

* Optional: give the displayed data cleaner formatting

### Calculation for District Summary (Totals, Averages, Percentages)

In [444]:
# Calculate the total number of schools
total_num_schools = school_data_complete['School ID'].nunique()

# Calculate the total number of students
total_num_students = school_data_complete['Student ID'].nunique()

# Calculate the total budget
total_budget = sum(school_data_complete['budget'].unique())

# Calculate the average 'math_score'
avg_math_score = school_data_complete['math_score'].mean()

# Calculate the average 'reading_score'
avg_reading_score = school_data_complete['reading_score'].mean()

# Calculate the percentage of students with a passing 'math_score' (70 or greater)
students_passing_math = school_data_complete.loc[school_data_complete['math_score']>=70].count()['student_name']
percent_students_passing_math= (students_passing_math/total_num_students)*100

                        
# Calculate the percentage of students with a passing 'reading_score' (70 or greater)
students_passing_reading = school_data_complete.loc[school_data_complete['reading_score']>=70].count()['student_name']
percent_students_passing_reading = (students_passing_reading/total_num_students)*100

# Calculate the overall of students who passed math and reading                             
overall_passing = school_data_complete.loc[(school_data_complete['reading_score'] >= 70) & 
                                           (school_data_complete['math_score'] >= 70)].count()['student_name']

# Calculate the percentage of students who passed math and reading (% Overall Passing)
percent_overall = (overall_passing / total_num_students) *100


# Create a dataframe to hold the above results
district_summary = pd.DataFrame({'Total Schools':[total_num_schools],
                                 'Total Students': [total_num_students],
                                 'Total Budget':[total_budget],
                                 'Average Math Score':[avg_math_score],
                                 'Average Reading Score': [avg_reading_score],
                                 'Passing Math': [percent_students_passing_math],
                                 'Passing Reading':[percent_students_passing_reading],
                                 'Overall Passing':[percent_overall]})


# Formatting for a cleaner DataFrame
district_summary ["Total Students"] = district_summary ["Total Students"].map("{:,}".format)
district_summary ["Total Budget"] = district_summary ["Total Budget"].map("${:,.0f}".format)
district_summary ["Average Math Score"] = district_summary ["Average Math Score"].map("{:,.2f}".format)
district_summary ["Average Reading Score"] = district_summary ["Average Reading Score"].map("{:,.2f}".format)
district_summary ["Passing Math"] = district_summary ["Passing Math"].map("{:,.2f} %".format)
district_summary ["Passing Reading"] = district_summary ["Passing Reading"].map("{:,.2f} %".format)
district_summary ["Overall Passing"] = district_summary ["Overall Passing"].map("{:,.2f} %".format)

# Display DataFrame
district_summary



Total Schools Total Students Total Budget Average Math Score  \
0             15         39,170  $24,649,428              78.99   

  Average Reading Score Passing Math Passing Reading Overall Passing  
0                 81.88      74.98 %         85.81 %         65.17 %

# School Summary

* Create an overview table that summarizes key metrics about each school, including:
  * School Name
  * School Type
  * Total Students
  * Total School Budget
  * Per Student Budget
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * % Overall Passing (The percentage of students that passed math **and** reading.)
  
* Create a dataframe to hold the above results

### Creating an overview table that summarizes key metrics about Each School

In [445]:
# First we need to groupby our table by School Name using school_data_complete
schools = school_data_complete.groupby(['school_name'])

# Get the school type for each school
school_type = schools['type'].first()

# Calculate Total Students for each school
total_students = schools['Student ID'].nunique()

# Calculate Total School Budget for each school
school_budget = schools['budget'].first()

# Calcualte Per Student Budget for each school
each_student_budget = school_budget / total_students

# Calculate Average Math Score for each school
avg_math_score = schools['math_score'].mean()

# Calculate Average Reading Score for each school
avg_reading_score = schools['reading_score'].mean()

# % # Calculate the percentange of passing math score for each school
# Used .size() since it returns size of dataframe/series which is equivalent to total number of elements.
student_passed_math = school_data_complete.loc[school_data_complete["math_score"] >= 70].groupby(['school_name']).size()
percent_pass_math = (student_passed_math/total_students)*100

# % # Calculate the percentange of passing reading score for each school
student_passed_reading = school_data_complete.loc[school_data_complete["reading_score"] >= 70].groupby(['school_name']).size()
percent_pass_reading = (student_passed_reading/total_students)*100


# % Overall Passing (The percentage of students that passed math and reading.) This part was Challenging!                              
overall_passing = school_data_complete[(school_data_complete['reading_score'] >= 70) & 
                                       (school_data_complete['math_score'] >= 70)].groupby('school_name')['Student ID'].count()
percent_overall = (overall_passing / total_students) *100


# Create a dataframe to hold the above results
school_summary = pd.DataFrame({"School Type":school_type,
                               "Total Students":total_students,
                               "Total School Budget": school_budget,
                               "Per Student Budget": each_student_budget,
                               "Avg Math Score": avg_math_score,
                               "Avg Reading Score":avg_reading_score,
                               "Passing Math %": percent_pass_math,
                               "Passing Reading %": percent_pass_reading,
                               "Overall Passing %": percent_overall})

# Create a copy of school summary data frame before formatting and turning it into strings so I will be able to use the numeric data on original data frame later
format_school_summary = school_summary.copy()

# Formatting for a cleaner DataFrame
format_school_summary ["Total School Budget"] = format_school_summary ["Total School Budget"].map("${:,.0f}".format)
format_school_summary ["Per Student Budget"] = format_school_summary ["Per Student Budget"].map("${:,.2f}".format)
format_school_summary ["Avg Math Score"] = format_school_summary["Avg Math Score"].map("{:,.2f}".format)
format_school_summary ["Avg Reading Score"] = format_school_summary ["Avg Reading Score"].map("{:,.2f}".format)
format_school_summary ["Passing Math %"] = format_school_summary ["Passing Math %"].map("{:,.2f} %".format)
format_school_summary ["Passing Reading %"] = format_school_summary ["Passing Reading %"].map("{:,.2f} %".format)
format_school_summary ["Overall Passing %"] = format_school_summary ["Overall Passing %"].map("{:,.2f} %".format)

# Display DataFrame withhout the index label
format_school_summary.index.name = None
format_school_summary

School Type  Total Students Total School Budget  \
Bailey High School       District            4976          $3,124,928   
Cabrera High School       Charter            1858          $1,081,356   
Figueroa High School     District            2949          $1,884,411   
Ford High School         District            2739          $1,763,916   
Griffin High School       Charter            1468            $917,500   
Hernandez High School    District            4635          $3,022,020   
Holden High School        Charter             427            $248,087   
Huang High School        District            2917          $1,910,635   
Johnson High School      District            4761          $3,094,650   
Pena High School          Charter             962            $585,858   
Rodriguez High School    District            3999          $2,547,363   
Shelton High School       Charter            1761          $1,056,600   
Thomas High School        Charter            1635          $1,043,130   
Wilson High School        Charter            2283          $1,319,574   
Wright High School        Charter            1800          $1,049,400   

                      Per Student Budget Avg Math Score Avg Reading Score  \
Bailey High School               $628.00          77.05             81.03   
Cabrera High School              $582.00          83.06             83.98   
Figueroa High School             $639.00          76.71             81.16   
Ford High School                 $644.00          77.10             80.75   
Griffin High School              $625.00          83.35             83.82   
Hernandez High School            $652.00          77.29             80.93   
Holden High School               $581.00          83.80             83.81   
Huang High School                $655.00          76.63             81.18   
Johnson High School              $650.00          77.07             80.97   
Pena High School                 $609.00          83.84             84.04   
Rodriguez High School            $637.00          76.84             80.74   
Shelton High School              $600.00          83.36             83.73   
Thomas High School               $638.00          83.42             83.85   
Wilson High School               $578.00          83.27             83.99   
Wright High School               $583.00          83.68             83.95   

                      Passing Math % Passing Reading % Overall Passing %  
Bailey High School           66.68 %           81.93 %           54.64 %  
Cabrera High School          94.13 %           97.04 %           91.33 %  
Figueroa High School         65.99 %           80.74 %           53.20 %  
Ford High School             68.31 %           79.30 %           54.29 %  
Griffin High School          93.39 %           97.14 %           90.60 %  
Hernandez High School        66.75 %           80.86 %           53.53 %  
Holden High School           92.51 %           96.25 %           89.23 %  
Huang High School            65.68 %           81.32 %           53.51 %  
Johnson High School          66.06 %           81.22 %           53.54 %  
Pena High School             94.59 %           95.95 %           90.54 %  
Rodriguez High School        66.37 %           80.22 %           52.99 %  
Shelton High School          93.87 %           95.85 %           89.89 %  
Thomas High School           93.27 %           97.31 %           90.95 %  
Wilson High School           93.87 %           96.54 %           90.58 %  
Wright High School           93.33 %           96.61 %           90.33 %

## Top Performing Schools (By % Overall Passing)

* Sort and display the top five performing schools by % overall passing.

In [446]:
# Sort and display the top five schools in overall passing rate, setting ascending to False to sort value from highest to lowest
top_performing_schools = school_summary.sort_values(by='Overall Passing %',ascending = False)

# Display DataFrame withhout the index label
top_performing_schools.index.name = None
top_performing_schools.head()

School Type  Total Students  Total School Budget  \
Cabrera High School     Charter            1858              1081356   
Thomas High School      Charter            1635              1043130   
Griffin High School     Charter            1468               917500   
Wilson High School      Charter            2283              1319574   
Pena High School        Charter             962               585858   

                     Per Student Budget  Avg Math Score  Avg Reading Score  \
Cabrera High School               582.0       83.061895          83.975780   
Thomas High School                638.0       83.418349          83.848930   
Griffin High School               625.0       83.351499          83.816757   
Wilson High School                578.0       83.274201          83.989488   
Pena High School                  609.0       83.839917          84.044699   

                     Passing Math %  Passing Reading %  Overall Passing %  
Cabrera High School       94.133477          97.039828          91.334769  
Thomas High School        93.272171          97.308869          90.948012  
Griffin High School       93.392371          97.138965          90.599455  
Wilson High School        93.867718          96.539641          90.582567  
Pena High School          94.594595          95.945946          90.540541

## Bottom Performing Schools (By % Overall Passing)

* Sort and display the five worst-performing schools by % overall passing.

In [447]:
# Sort and display the bottom five schools in overall passing rate, by default ascending is True, which will sort from lowest to highest
top_performing_schools = school_summary.sort_values(by='Overall Passing %')

# Display DataFrame withhout the index label
top_performing_schools.index.name = None
top_performing_schools.head()

School Type  Total Students  Total School Budget  \
Rodriguez High School    District            3999              2547363   
Figueroa High School     District            2949              1884411   
Huang High School        District            2917              1910635   
Hernandez High School    District            4635              3022020   
Johnson High School      District            4761              3094650   

                       Per Student Budget  Avg Math Score  Avg Reading Score  \
Rodriguez High School               637.0       76.842711          80.744686   
Figueroa High School                639.0       76.711767          81.158020   
Huang High School                   655.0       76.629414          81.182722   
Hernandez High School               652.0       77.289752          80.934412   
Johnson High School                 650.0       77.072464          80.966394   

                       Passing Math %  Passing Reading %  Overall Passing %  
Rodriguez High School       66.366592          80.220055          52.988247  
Figueroa High School        65.988471          80.739234          53.204476  
Huang High School           65.683922          81.316421          53.513884  
Hernandez High School       66.752967          80.862999          53.527508  
Johnson High School         66.057551          81.222432          53.539172

## Math Scores by Grade

* Create a table that lists the average Math Score for students of each grade level (9th, 10th, 11th, 12th) at each school.

* Create a pandas series for each grade. Hint: use a conditional statement.

* Group each series by school

* Combine the series into a dataframe

* Optional: give the displayed data cleaner formatting

In [448]:
# Calculate the average Math score for students of 9th, 10th, 11th, 12th grade at each school
grade_9 = school_data_complete.loc[school_data_complete['grade'] == '9th'].groupby(["school_name"])['math_score'].mean()                          
grade_10 = school_data_complete.loc[school_data_complete['grade'] == '10th'].groupby(["school_name"])['math_score'].mean()
grade_11 = school_data_complete.loc[school_data_complete['grade'] == '11th'].groupby(["school_name"])['math_score'].mean()
grade_12 = school_data_complete.loc[school_data_complete['grade'] == '12th'].groupby(["school_name"])['math_score'].mean()

# Create a dataframe to hold the above results 
scores_by_grade = pd.DataFrame({"9th Grade Avg Math Scores":grade_9,
                               "10th Grade Avg Math Scores":grade_10,
                               "11th Grade Avg Math Scores":grade_11,
                               "12th Grade Avg Math Scores":grade_12})

# Display DataFrame withhout the index label
scores_by_grade.index.name = None
scores_by_grade

9th Grade Avg Math Scores  10th Grade Avg Math Scores  \
Bailey High School                     77.083676                   76.996772   
Cabrera High School                    83.094697                   83.154506   
Figueroa High School                   76.403037                   76.539974   
Ford High School                       77.361345                   77.672316   
Griffin High School                    82.044010                   84.229064   
Hernandez High School                  77.438495                   77.337408   
Holden High School                     83.787402                   83.429825   
Huang High School                      77.027251                   75.908735   
Johnson High School                    77.187857                   76.691117   
Pena High School                       83.625455                   83.372000   
Rodriguez High School                  76.859966                   76.612500   
Shelton High School                    83.420755                   82.917411   
Thomas High School                     83.590022                   83.087886   
Wilson High School                     83.085578                   83.724422   
Wright High School                     83.264706                   84.010288   

                       11th Grade Avg Math Scores  12th Grade Avg Math Scores  
Bailey High School                      77.515588                   76.492218  
Cabrera High School                     82.765560                   83.277487  
Figueroa High School                    76.884344                   77.151369  
Ford High School                        76.918058                   76.179963  
Griffin High School                     83.842105                   83.356164  
Hernandez High School                   77.136029                   77.186567  
Holden High School                      85.000000                   82.855422  
Huang High School                       76.446602                   77.225641  
Johnson High School                     77.491653                   76.863248  
Pena High School                        84.328125                   84.121547  
Rodriguez High School                   76.395626                   77.690748  
Shelton High School                     83.383495                   83.778976  
Thomas High School                      83.498795                   83.497041  
Wilson High School                      83.195326                   83.035794  
Wright High School                      83.836782                   83.644986

## Reading Score by Grade 

* Create a table that lists the average Reading Score for students of each grade level (9th, 10th, 11th, 12th) at each school.

* Create a pandas series for each grade.

* Group each series by school

* Combine the series into a dataframe

* Optional: give the displayed data cleaner formatting

In [449]:
# Calculate the average Reading score for students of 9th, 10th, 11th, 12th grade at each school
grade_9 = school_data_complete.loc[school_data_complete['grade'] == '9th'].groupby(["school_name"])['reading_score'].mean()                          
grade_10 = school_data_complete.loc[school_data_complete['grade'] == '10th'].groupby(["school_name"])['reading_score'].mean()
grade_11 = school_data_complete.loc[school_data_complete['grade'] == '11th'].groupby(["school_name"])['reading_score'].mean()
grade_12 = school_data_complete.loc[school_data_complete['grade'] == '12th'].groupby(["school_name"])['reading_score'].mean()

# Create a dataframe to hold the above results 
scores_by_grade = pd.DataFrame({"9th Grade Avg Reading Scores":grade_9,
                               "10th Grade Avg Reading Scores":grade_10,
                               "11th Grade Avg Reading Scores":grade_11,
                               "12th Grade Avg Reading Scores":grade_12})

# Display DataFrame withhout the index label
scores_by_grade.index.name = None
scores_by_grade

9th Grade Avg Reading Scores  \
Bailey High School                        81.303155   
Cabrera High School                       83.676136   
Figueroa High School                      81.198598   
Ford High School                          80.632653   
Griffin High School                       83.369193   
Hernandez High School                     80.866860   
Holden High School                        83.677165   
Huang High School                         81.290284   
Johnson High School                       81.260714   
Pena High School                          83.807273   
Rodriguez High School                     80.993127   
Shelton High School                       84.122642   
Thomas High School                        83.728850   
Wilson High School                        83.939778   
Wright High School                        83.833333   

                       10th Grade Avg Reading Scores  \
Bailey High School                         80.907183   
Cabrera High School                        84.253219   
Figueroa High School                       81.408912   
Ford High School                           81.262712   
Griffin High School                        83.706897   
Hernandez High School                      80.660147   
Holden High School                         83.324561   
Huang High School                          81.512386   
Johnson High School                        80.773431   
Pena High School                           83.612000   
Rodriguez High School                      80.629808   
Shelton High School                        83.441964   
Thomas High School                         84.254157   
Wilson High School                         84.021452   
Wright High School                         83.812757   

                       11th Grade Avg Reading Scores  \
Bailey High School                         80.945643   
Cabrera High School                        83.788382   
Figueroa High School                       80.640339   
Ford High School                           80.403642   
Griffin High School                        84.288089   
Hernandez High School                      81.396140   
Holden High School                         83.815534   
Huang High School                          81.417476   
Johnson High School                        80.616027   
Pena High School                           84.335938   
Rodriguez High School                      80.864811   
Shelton High School                        84.373786   
Thomas High School                         83.585542   
Wilson High School                         83.764608   
Wright High School                         84.156322   

                       12th Grade Avg Reading Scores  
Bailey High School                         80.912451  
Cabrera High School                        84.287958  
Figueroa High School                       81.384863  
Ford High School                           80.662338  
Griffin High School                        84.013699  
Hernandez High School                      80.857143  
Holden High School                         84.698795  
Huang High School                          80.305983  
Johnson High School                        81.227564  
Pena High School                           84.591160  
Rodriguez High School                      80.376426  
Shelton High School                        82.781671  
Thomas High School                         83.831361  
Wilson High School                         84.317673  
Wright High School                         84.073171

## Scores by School Spending

* Create a table that breaks down school performances based on average Spending Ranges (Per Student). Use 4 reasonable bins to group school spending. Include in the table each of the following:
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * Overall Passing Rate (Average of the above two)

### Figure out the Min and Max budget 

In [450]:
# Calculate the minimum and maximum budget for school_summary
print(school_summary['Per Student Budget'].min())
print(school_summary['Per Student Budget'].max())

578.0
655.0


In [451]:
# Create bins in which to place values based upon Per Student Budget
bins = [0, 585, 615, 645, 675]

# Create labels for these bins
spending_labels = ["<$585", "$585-615", "$615-645", "$645-675"]

# Slice the data and add a new column named 'Spending Ranges (Per Student)'
school_summary["Spending Ranges (Per Student)"] = pd.cut(school_summary["Per Student Budget"], bins, labels=spending_labels)

# Create a GroupBy object based upon "Spending Ranges (Per Student)"
group_by_spending = school_summary.groupby("Spending Ranges (Per Student)")

# Get the average of each desired column within the GroupBy spending
scores_by_spending = group_by_spending [["Avg Math Score", "Avg Reading Score", "Passing Math %", "Passing Reading %", "Overall Passing %"]].mean()

# Formatting for a cleaner DataFrame
scores_by_spending ["Passing Math %"] = scores_by_spending ["Passing Math %"].map("{:,.2f} %".format)
scores_by_spending ["Passing Reading %"] = scores_by_spending ["Passing Reading %"].map("{:,.2f} %".format)
scores_by_spending["Overall Passing %"] =scores_by_spending["Overall Passing %"].map("{:,.2f} %".format)

scores_by_spending.head()


##-----------------------------------------------------------------------------##
#        **** Below is a diffrent/longer way of doing this section****
##-----------------------------------------------------------------------------##

# # Create a new data frame from 'copied school_summary dataframe' by locating the desired columns
# scores_by_spending = school_summary.loc[:,["Avg Mat"Passing Math %"h Score",
#                                          "Avg Reading Score",
#                                          "Passing Math %",
#                                          "Passing Reading %",
#                                          "Overall Passing %",]]

# # Place the data series into a new column inside of the new DataFrame created 'scores_by_spending'
# scores_by_spending["Spending Ranges (Per Student)"] = pd.cut(school_summary["Per Student Budget"], bins, labels= spending_labels)

# # Create a group based off of the bins
# scores_by_spending = scores_by_spending.groupby('Spending Ranges (Per Student)').mean()

## Formatting for a cleaner DataFrame
# scores_by_spending ["Passing Math %"] = scores_by_spending ["Passing Math %"].map("{:,.2f} %".format)
# scores_by_spending ["Passing Reading %"] = scores_by_spending ["Passing Reading %"].map("{:,.2f} %".format)
# scores_by_spending["Overall Passing %"] =scores_by_spending["Overall Passing %"].map("{:,.2f} %".format)

# scores_by_spending.head()

Avg Math Score  Avg Reading Score  \
Spending Ranges (Per Student)                                      
<$585                               83.455399          83.933814   
$585-615                            83.599686          83.885211   
$615-645                            79.079225          81.891436   
$645-675                            76.997210          81.027843   

                              Passing Math % Passing Reading %  \
Spending Ranges (Per Student)                                    
<$585                                93.46 %           96.61 %   
$585-615                             94.23 %           95.90 %   
$615-645                             75.67 %           86.11 %   
$645-675                             66.16 %           81.13 %   

                              Overall Passing %  
Spending Ranges (Per Student)                    
<$585                                   90.37 %  
$585-615                                90.22 %  
$615-645                                66.11 %  
$645-675                                53.53 %

## Scores by School Size

* Create a table that breaks down school performances based on School Size. Use 4 reasonable bins to group school size. 

* Include in the table each of the following:
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * Overall Passing Rate (Average of the above two)

In [452]:
print(school_summary['Total Students'].min())
print(school_summary['Total Students'].max())

427
4976


In [453]:
# Create bins in which to place values based upon Total Students
bins = [0,1000, 2000, 5000]

# Create labels for these bins
size_labels = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

# Slice the data and add a new column named 'School Size'
school_summary["School Size"] = pd.cut(school_summary["Total Students"], bins, labels=size_labels)

# Create a GroupBy object based upon "School Size"
group_by_size = school_summary.groupby("School Size")

# Get the average of each desired column within the GroupBy size
scores_by_size = group_by_size [["Avg Math Score", "Avg Reading Score", "Passing Math %", "Passing Reading %", "Overall Passing %"]].mean()

# Formatting for a cleaner DataFrame
scores_by_size["Passing Math %"] = scores_by_size ["Passing Math %"].map("{:,.2f} %".format)
scores_by_size ["Passing Reading %"] = scores_by_size ["Passing Reading %"].map("{:,.2f} %".format)
scores_by_size["Overall Passing %"] =scores_by_size["Overall Passing %"].map("{:,.2f} %".format)

scores_by_size.head()

Avg Math Score  Avg Reading Score Passing Math %  \
School Size                                                            
Small (<1000)            83.821598          83.929843        93.55 %   
Medium (1000-2000)       83.374684          83.864438        93.60 %   
Large (2000-5000)        77.746417          81.344493        69.96 %   

                   Passing Reading % Overall Passing %  
School Size                                             
Small (<1000)                96.10 %           89.88 %  
Medium (1000-2000)           96.79 %           90.62 %  
Large (2000-5000)            82.77 %           58.29 %

## Scores by School Type

* Create a table that breaks down school performances based on School Type

* Include in the table each of the following:
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * Overall Passing Rate (Average of the above two)

In [454]:
# Create a GroupBy object based upon "School Type"
group_by_size = school_summary.groupby("School Type")

# Get the average of each desired column within the GroupBy school type
scores_by_size = group_by_size [["Avg Math Score", "Avg Reading Score", "Passing Math %", "Passing Reading %", "Overall Passing %"]].mean()

# Formatting for a cleaner DataFrame
scores_by_size["Passing Math %"] = scores_by_size ["Passing Math %"].map("{:,.2f} %".format)
scores_by_size ["Passing Reading %"] = scores_by_size ["Passing Reading %"].map("{:,.2f} %".format)
scores_by_size["Overall Passing %"] =scores_by_size["Overall Passing %"].map("{:,.2f} %".format)

scores_by_size.head()

Avg Math Score  Avg Reading Score Passing Math %  \
School Type                                                     
Charter           83.473852          83.896421        93.62 %   
District          76.956733          80.966636        66.55 %   

            Passing Reading % Overall Passing %  
School Type                                      
Charter               96.59 %           90.43 %  
District              80.80 %           53.67 %

## PyCity Schools Analysis

* As a whole, schools with higher budgets, did not yield better test results. By contrast, schools with higher spending per student actually (\\$645- \\$675) underperformed compared to schools with smaller budgets (\$585 per student).

* As a whole, smaller and medium sized schools dramatically out-performed large sized schools on passing math performances (89-91% passing vs 67%).

* As a whole, charter schools out-performed the public district schools across all metrics. However, more analysis will be required to glean if the effect is due to school practices or the fact that charter schools tend to serve smaller student populations per school.